application/comit_node/src/connection_pool.rs

```diff
@@ -0,0 +1,286 @@
+                    node_id
+                );
+            }
+            // The most common case. We were disconnected but now we're connected.
+            ConnectionState::Disconnected => {
+                Self::transition_to_connected(
+                    node_id,
+                    &mut *existing_connection_guard,
+                    Arc::clone(&existing_connection),
+                    bob_spawner,
+                    socket,
+                );
+            }
+        }
+    }
+
+    /// Sets the connection_state to "Connected" and spawns the Connection.
+    /// When the connection is finished it sets it to Disconnected.
+    /// `connection_state` and `connection_handle`` must point to the same
+    /// ConnectionState.
+    fn transition_to_connected<S: AsyncRead + AsyncWrite + Send + 'static, B: BobSpawner>(
+        node_id: NodeId,
+        connection_state: &mut ConnectionState,
+        connection_handle: Arc<Mutex<ConnectionState>>,
+        bob_spawner: B,
+        socket: S,
+    ) -> BamClient {
+        info!(
+            "Established new connection to {} while connection was in the {} state",
+            node_id, connection_state
+        );
+
+        let (client, connection) = {
+            // Set up BAM connection
+            let codec = json::JsonFrameCodec::default();
+            let response_source = Arc::new(Mutex::new(json::JsonResponseSource::default()));
+            let incoming_frames = json::JsonFrameHandler::create(
+                swap_config(bob_spawner),
+                Arc::clone(&response_source),
+            );
+            let (client, outgoing_frames) = bam::client::Client::create(response_source);
+
+            (
+                client,
+                bam::connection::new(codec, socket, incoming_frames, outgoing_frames),
+            )
+        };
+
+        let client = Arc::new(Mutex::new(client));
+
+        *connection_state = ConnectionState::Connected {
+            client: Arc::clone(&client),
+        };
+
+        let final_connection_state = Arc::clone(&connection_handle);
+
+        tokio::spawn(connection.then(move |result| {
+            let mut final_connection_state = final_connection_state.lock().unwrap();
+
+            match result {
+                Ok(_) => {
+                    info!(
+                        "Connection to {} was closed while it was in the {} state",
+                        node_id, final_connection_state
+                    );
+                }
+                Err(e) => {
+                    error!(
+                        "Connection to {} prematurely closed while it was in the {} state: {:?}",
+                        node_id, final_connection_state, e
+                    );
+                }
+            }
+
+            *final_connection_state = ConnectionState::Disconnected;
+
+            Ok(())
+        }));
+
+        client
+    }
+
+    pub fn connected_addrs(&self) -> Vec<SocketAddr> {
```

Comment:
:question: Don't you need to filter by `ConnectedState::Connected` here?

Reply:
Yes. But I'd rather write tests for this new feature before implementing it.

Reply:
Should we remove the function then? :D